plugins/BEdita/API/tests/TestCase/Middleware/TokenMiddlewareTest.php

```diff
@@ -0,0 +1,231 @@
+                    'HTTP_X_API_KEY' => 'this API key is invalid!',
+                ],
+            ],
+            'invalid query API key' => [
+                new ForbiddenException('Invalid API key'),
+                [],
+                [
+                    'api_key' => 'gustavo',
+                ],
+            ],
+            'missing API key' => [
+                new ForbiddenException('Missing API key'),
+                [],
+                [],
+                true,
+            ],
+            'api API query' => [
+                [
+                    'app' => ['id' => 1],
+                    'payload' => null,
+                ],
+                [],
+                [
+                    'api_key' => API_KEY,
+                ],
+            ],
+            'no token' => [
+                [
+                    'payload' => null,
+                ],
+                [
+                ],
+            ],
+        ];
+    }
+
+    /**
+     * Test middleware invocation.
+     *
+     * @return void
+     *
+     * @dataProvider invokeProvider
+     *
+     * @covers ::__invoke()
+     * @covers ::readApplication()
+     * @covers ::getToken()
+     * @covers ::applicationFromApiKey()
+     * @covers ::fetchApiKey()
+     * @covers ::verifyClientCredentials()
+     */
+    public function testInvoke($expected, array $server, ?array $query = null, bool $blockAnonymous = false): void
+    {
+        if ($expected instanceof \Exception) {
+            $this->expectException(get_class($expected));
+            $this->expectExceptionMessage($expected->getMessage());
+        }
+        Configure::write('Security.blockAnonymousApps', $blockAnonymous);
+
+        $request = ServerRequestFactory::fromGlobals($server, $query);
+        $response = new Response();
+        $next = function ($req, $res) {
+            return $req;
+        };
+        $middleware = new TokenMiddleware();
+        /** @var \Zend\Diactoros\ServerRequest $result */
+        $result = $middleware($request, $response, $next);
+
+        $app = CurrentApplication::getApplication();
+        if (empty($expected['app'])) {
+            static::assertNull($app);
+        } else {
+            $id = Hash::get($expected, 'app.id');
+            static::assertEquals($app->id, $id);
+        }
+
+        $payload = $result->getAttribute(TokenMiddleware::PAYLOAD_REQUEST_ATTRIBUTE, null);
+        $expectedPayload = Hash::get($expected, 'payload');
+        static::assertEquals($expectedPayload, $payload);
+    }
+
+    /**
+     * Test default behavior on missing 'Security.blockAnonymousApps' key
+     *
+     * @return void
+     *
+     * @covers ::fetchApiKey()
+     * @covers ::verifyClientCredentials()
+     */
+    public function testGetApplicationDefault()
+    {
+        $this->expectException(ForbiddenException::class);
+        $this->expectExceptionMessage('Missing API key');
+
+        Configure::delete('Security.blockAnonymousApps');
```

Comment:
Do we need to restore previously value of `'Security.blockAnonymousApps'` at the end of test?

Reply:
In previous tests we did not keep state of this configuration. We can assume it's deleted or not defined as default, 
Here we just make sure it is deleted.